tools/clang/lib/AST/ASTContextHLSL.cpp

```diff
@@ -938,11 +938,9 @@ CXXRecordDecl *hlsl::DeclareTemplateTypeWithHandleInDeclContext(
   QualType elementType = context.getTemplateTypeParmType(
       /*templateDepth*/ 0, 0, ParameterPackFalse, elementTemplateParamDecl);
 
-  if (templateArgCount > 1 &&
-      // Only need array type for inputpatch and outputpatch.
-      // Avoid Texture2DMS which may use 0 count.
-      // TODO: use hlsl types to do the check.
-      !name.startswith("Texture") && !name.startswith("RWTexture")) {
+  if (templateArgCount > 1 && (!Attr || !isa<HLSLTypedResourceAttr>(Attr))) {
+    // Only need array type for inputpatch and outputpatch.
+    // isTyped check avoids Texture2DMS which may use 0 count.
```

Comment:
This is super ugly, assuming that we can determine whether this is an InputPatch or OutputPatch based on whether it has two or more template arguments and is not a typed resource.

In PR #7143, you add the `HLSLTessPatch` attribute [here](https://github.com/microsoft/DirectXShaderCompiler/pull/7143/files#diff-ece93d307d53026d6a66bb7372fe5af375c97df5c5913a10bf591d61ae3a3bcaR1001).

It would be much cleaner to depend on that attribute for positive identification instead of what it does now.

Could the required attributes be added in an earlier NFC change that these PRs depend on?

Reply:
In fact, the refactor of these functions to accept and apply the attribute could be included with this earlier NFC change.